text/0059-otlp-trace-data-format.md

```diff
@@ -0,0 +1,346 @@
+    bytes span_id = 2;
+
+    // tracestate is the trace state associated with the link.
+    repeated TraceStateEntry tracestate = 3;
+
+    // attributes is a collection of attribute key/value pairs on the link.
+    repeated AttributeKeyValue attributes = 4;
+
+    // dropped_attributes_count is the number of dropped attributes. If the value is 0,
+    // then no attributes were dropped.
+    int32 dropped_attributes_count = 5;
+  }
+
+  // links is a collection of Links, which are references from this span to a span
+  // in the same or different trace.
+  repeated Link links = 14;
+
+  // dropped_links_count is the number of dropped links after the maximum size was
+  // enforced. If this value is 0, then no links were dropped.
+  int32 dropped_links_count = 15;
+
+  // status is an optional final status for this span. Semantically when status
+  // wasn't set it is means span ended without errors and assume Status.Ok (code = 0).
+  Status status = 16;
+
+  // child_span_count is an optional number of child spans that were generated while this
+  // span was active. If set, allows an implementation to detect missing child spans.
+  int32 child_span_count = 17;
+}
+
+// The `Status` type defines a logical error model that is suitable for different
+// programming environments, including REST APIs and RPC APIs.
+message Status {
+  // The status code. This is optional field. It is safe to assume 0 (OK)
+  // when not set.
+  int32 code = 1;
+
+  // A developer-facing error message, which should be in English.
+  string message = 2;
+}
+```
+
+### AttributeKeyValue
+
+```protobuf
+// AttributeKeyValue is a key-value pair that is used to store Span attributes, Resource
+// labels, etc.
+message AttributeKeyValue {
+  // ValueType is the enumeration of possible types that value can have.
+  enum ValueType {
+    STRING  = 0;
+    BOOL    = 1;
+    INT64   = 2;
+    DOUBLE  = 3;
+  };
+
+  // key part of the key-value pair.
+  string key = 1;
+
+  // The type of the value.
+  ValueType type = 2;
+
+  // Only one of the following fields is supposed to contain data (determined by `type` field value).
+  // This is deliberately not using Protobuf `oneof` for performance reasons (verified by benchmarks).
+
+  // A string up to 256 bytes long.
+  string string_value = 3;
+  // A 64-bit signed integer.
+  int64 int64_value = 4;
+  // A Boolean value represented by `true` or `false`.
+  bool bool_value = 5;
+  // A double value.
+  double double_value = 6;
+}
+```
+
+## Trade-offs and mitigations
+
+Timestamps were changed from google.protobuf.Timestamp to a int64 representation in Unix epoch nanoseconds. This change reduces the type-safety but benchmarks show that for small spans there is 15-20% encoding/decoding CPU speed gain. This is the right trade-off to make because encoding/decoding CPU consumption tends to dominate many workloads (particularly in OpenTelemetry Service).
```

Comment:
wow. I wonder than why anybody uses `google.protobuf.Timestamp`?

Reply:
1. google.protobuf.Timestamp supports larger date range, but I think we are fine with supporting only for the next 240 years in this protocol. This is plenty of time for our successors to come up with a better protocol :-)
2. google.protobuf.Timestamp value also can be omitted, unspecified, while timestamp in nanoseconds cannot (if you omit it you will get 0 value which is a valid timestamp - so cannot differentiate a valid 0 timestamp from missing timestamp). I believe this is also a non-issue for traces since trace timestamps are mandatory.